types/src/sample.rs

```diff
@@ -2,45 +2,153 @@ use std::mem::size_of;
 
 use blockstore::block::CidError;
 use bytes::{BufMut, BytesMut};
+use celestia_proto::proof::pb::Proof as RawProof;
+use celestia_proto::share::p2p::shwap::Sample as RawSample;
 use cid::CidGeneric;
 use multihash::Multihash;
+use nmt_rs::nmt_proof::NamespaceProof as NmtNamespaceProof;
+use nmt_rs::NamespaceMerkleHasher;
+use serde::{Deserialize, Serialize};
+use tendermint::Hash;
+use tendermint_proto::Protobuf;
 
-use crate::axis::{AxisId, AxisType};
-use crate::DataAvailabilityHeader;
-use crate::{Error, Result};
+use crate::extended_data_square::{AxisType, ExtendedDataSquare};
+use crate::nmt::{NamespaceProof, NamespacedHash, NamespacedHashExt, NamespacedSha2Hasher, Nmt};
+use crate::row::RowId;
+use crate::{Error, Result, Share};
 
 const SAMPLE_ID_SIZE: usize = SampleId::size();
 pub const SAMPLE_ID_MULTIHASH_CODE: u64 = 0x7801;
 pub const SAMPLE_ID_CODEC: u64 = 0x7800;
 
-/// Represents particular sample along the axis on specific Data Square
+/// Represents a location of a sample on the EDS
 #[derive(Debug, PartialEq, Clone, Copy)]
 pub struct SampleId {
-    pub axis: AxisId,
+    pub row: RowId,
     pub index: u16,
 }
 
-impl SampleId {
-    /// Create new SampleId. Index references sample number from the entire Data Square (is
-    /// converted to row/col coordinates internally). Same location can be sampled row or
-    /// column-wise, axis_type is used to distinguish that. Axis root hash is calculated from the
-    /// DataAvailabilityHeader
-    #[allow(dead_code)] // unused for now
+/// Represents Sample, with proof of its inclusion and location on EDS
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(try_from = "RawSample", into = "RawSample")]
+pub struct Sample {
+    pub sample_id: SampleId,
+
+    pub sample_proof_type: AxisType,
+    pub share: Share,
+    pub proof: NamespaceProof,
+}
+
+impl Sample {
+    /// Create a new Sample. Index references sample number from the entire Data Square
     pub fn new(
         axis_type: AxisType,
         index: usize,
```

Comment:
I'm unsure about this `index` + `axis` api. Is there a reason for it to be more convenient than providing the `x` and `y`?

Reply:
if you decide to change that, please update also in SampleId

Reply:
`axis` isn't used to specify the location of the sample, but only to indicate whether proof is calculated by row or column. Notice how `SampleId` doesn't  use `AxisType` at all, but always works in row-major order.
As for the `index` vs pair `x` + `y` itself, I don't have strong opinions here as I've not yet seen our usage of it, so I copied the API from celestia :P

Reply:
hmmm I think they use it only in the DASer and sampling index for a one dimensional array is simpler as you sample only one number

Reply:
I'm concerned because I guess celestia guys have it as something pretty internal somewhere near shwap p2p protocol, whereas we expose it from a pub api of some crate. I'd either:
- switch to x, y always row-major in the `new` and rename this one as let's say `with_axis`
- move the shwap types entirely into lumina-node and hide them

Reply:
@oblique wdyt?

Reply:
since oblique has different topics currently, I think it's fine to go with the current one and leave a todo whether we would want an x, y api in a future